### PR TITLE
[CH] support months_between function

### PR DESCRIPTION
**Pull request by zhanglistar**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

(Fixes: \#2348)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
UT

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: cpp-ch/local-engine/Parser/scalar_function_parser/months_between.cpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Function {} requires at least two arguments", getName());
+
+        const auto * round_arg = parsed_args.size() >= 3 ? parsed_args[2] : nullptr;
+        const auto * timezone_arg = parsed_args.size() >= 4 ? parsed_args[3] : nullptr;
+        const auto * unit_const_node = addColumnToActionsDAG(actions_dag, std::make_shared<DataTypeString>(), "day");
+        const auto * magic_const_node = addColumnToActionsDAG(actions_dag, std::make_shared<DataTypeFloat32>(), 30.636038795303726);
+        const auto * date_diff_node = toFunctionNode(actions_dag, "date_diff", {unit_const_node, parsed_args[1], parsed_args[0], timezone_arg});
+        const auto * devide_node = toFunctionNode(actions_dag, "divide", {date_diff_node, magic_const_node});
+        const auto * round_node = toFunctionNode(actions_dag, "round", {devide_node, addColumnToActionsDAG(actions_dag, std::make_shared<DataTypeInt32>(), 8)});
+        return toFunctionNode(actions_dag, "if", {round_arg, round_node, devide_node});
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/scalar_function_parser/months_between.cpp:50]: return加一个convertNodeTypeIfNeeded，避免输出类型与substrait function的output_type不一致

Comment (taiyang-li) [reply cpp-ch/local-engine/Parser/scalar_function_parser/months_between.cpp:50]: 参考cpp-ch/local-engine/Parser/example_udf/myMd5.cpp

File: gluten-core/src/main/scala/io/glutenproject/expression/DateTimeExpressionsTransformer.scala [hunk 2]
```diff
@@ -198,6 +198,45 @@ class TruncTimestampTransformer(
...
+      Seq(original.date1.dataType, original.date2.dataType, original.roundOff.dataType, StringType)
+    } else {
+      Seq(original.date1.dataType, original.date2.dataType, original.roundOff.dataType)
+    }
+
+    val functionId = ExpressionBuilder.newScalarFunction(functionMap,
+      ConverterUtils.makeFuncName(substraitExprName, dataTypes))
+
+    val expressionNodes = new java.util.ArrayList[ExpressionNode]()
+    // val lowerFormatNode = ExpressionBuilder.makeStringLiteral("day")
```
Comment (taiyang-li) [inline gluten-core/src/main/scala/io/glutenproject/expression/DateTimeExpressionsTransformer.scala:226]: remove useless codes

File: backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseDatetimeExpressionSuite.scala [hunk 3]
```diff
@@ -148,4 +148,19 @@ class GlutenClickHouseDatetimeExpressionSuite
       compareResultsAgainstVanillaSpark(sql2, true, { _ => }, false)
     }
   }
+
+  test("test months_between function") {
```
Comment (taiyang-li) [inline backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseDatetimeExpressionSuite.scala:152]: months_between有没有spark官方的ut, 可以在ClickhouseTestSettings里开启。

**Review by taiyang-li — CHANGES_REQUESTED** [review]

File: backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseDatetimeExpressionSuite.scala (hunk 3 above)
Comment (zhanglistar) [reply backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseDatetimeExpressionSuite.scala:152]: 现在有精度问题， 还需要进一步缩小精度。spark是到微妙，但是ch只有到秒。

**Comment by zhanglistar** [comment]
Now we have precision problem, spark support micro seconds but ch only seconds. Need to work.

**Review by zzcclp — APPROVED** [review]
LGTM

**Review by liuneng1994 — APPROVED** [review]
LGTM

**Pull request merged**